### PR TITLE
unix time module now return result

### LINKED ISSUE #108277: std::fs::Metadata timestamp methods will panic on files with invalid timestamp nsec values
`std::fs::Metadata` timestamp methods (like `created`) will panic on files with invalid timestamp nsec values.`

It will panic here:
`thread 'async-std/runtime' panicked at 'assertion failed: tv_nsec >= 0 && tv_nsec < NSEC_PER_SEC as i64', library/std/src/sys/unix/time.rs:66:9`

We have recently [stumbled upon](https://gitlab.gnome.org/World/pika-backup/-/issues/328#note_1678169) multiple btrfs users that happen to have files that look something like this:

```
# stat ./.temp-FOYNU1
File: ./.temp-FOYNU1
Size: 37271     	Blocks: 80         IO Block: 4096   regular file
Device: 0,40	Inode: 4024244     Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/   x)   Gid: ( 1000/   x)
Context: unconfined_u:object_r:gkeyringd_gnome_home_t:s0
Access: 2023-02-19 21:35:19.087361106 +0100
Modify: 2022-10-29 09:48:58.344737238 +0200
Change: 2022-10-29 09:48:58.344737238 +0200
Birth: 5006491383123099853.-1424152917
```

This is most likely a filesystem / kernel bug. But these files / broken kernels exist now and will cause any access to `ctime` (and possible the other time values) to panic.

My proposal would be to check bounds of the underlying syscall results and return an appropriate `Err` for these broken time values if the ctime / mtime / atime etc. contains values that would otherwise trigger this assertion.

### Meta
`rustc --version --verbose`:

```
rustc 1.67.1 (d5a82bbd2 2023-02-07)
binary: rustc
commit-hash: d5a82bbd26e1ad8b7401f6a718a9c57c96905483
commit-date: 2023-02-07
host: x86_64-unknown-linux-gnu
release: 1.67.1
LLVM version: 15.0.6
```

<details><summary>Backtrace</summary>
<p>

```
   0:     0x556e2422048a - std::backtrace_rs::backtrace::libunwind::trace::h79937bc171ada62c
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x556e2422048a - std::backtrace_rs::backtrace::trace_unsynchronized::h2292bca8571cb919
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x556e2422048a - std::sys_common::backtrace::_print_fmt::h9c461f248e4ae90d
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x556e2422048a - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::he9fe6bf1a39182e1
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x556e2424687e - core::fmt::write::h032658c119c720d7
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/fmt/mod.rs:1208:17
   5:     0x556e24219c65 - std::io::Write::write_fmt::h299fc90dfae41c0d
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/io/mod.rs:1682:15
   6:     0x556e24220255 - std::sys_common::backtrace::_print::heb70d25df9937e3f
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x556e24220255 - std::sys_common::backtrace::print::had745c0a76b8b521
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x556e24221cff - std::panicking::default_hook::{{closure}}::h1ea782cdfa2fd097
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:267:22
   9:     0x556e24221a3b - std::panicking::default_hook::h1cc3af63455a163c
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:286:9
  10:     0x556e2422240c - std::panicking::rust_panic_with_hook::h5cafdc4b3bfd5528
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:688:13
  11:     0x556e24222162 - std::panicking::begin_panic_handler::{{closure}}::hf31c60f40775892c
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:577:13
  12:     0x556e2422093c - std::sys_common::backtrace::__rust_end_short_backtrace::h28a5c7be595826cd
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:137:18
  13:     0x556e24221eb2 - rust_begin_unwind
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
  14:     0x556e237c2243 - core::panicking::panic_fmt::h8fa27a0b37dd98b7
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
  15:     0x556e237c231d - core::panicking::panic::h545818946343732b
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:111:5
  16:     0x556e2421704d - std::sys::unix::time::Timespec::new::h0e233601cbadebfd
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys/unix/time.rs:66:9
  17:     0x556e2421704d - std::sys::unix::time::SystemTime::new::h9d5242895ff4c7e2
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys/unix/time.rs:29:25
  18:     0x556e2421704d - std::sys::unix::fs::FileAttr::created::hb18250e52af180d4
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys/unix/fs.rs:485:24
  19:     0x556e2421704d - std::fs::Metadata::created::h80fcc7cb446092ab
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/fs.rs:1311:9
```

</p>
</details>

### PR DESCRIPTION
**Pull request by Stargateur**
First try to fix #108277 without break anything.

if anyone who read this know tips to be able to check compilation for different target I could use some help. So far I installed many target with rustup but `./x check --all-targets` doesn't seem to use them.

TODO:

- [x] better error
- [ ] test, how ?

@rustbot label -S-waiting-on-author +S-waiting-on-review

### CODE REVIEW

**Comment by thomcc** [comment]
> if anyone who read this know tips to be able to check compilation for different target I could use some help. So far I installed many target with rustup but ./x check --all-targets doesn't seem to use them.

I don't think `--all-targets` is right, I think you need to check each `--target` individually? Note that it does need to be `check`, and you may need to ensure the llvm submodule is *not* checked out (this may have been fixed).

**Comment by Stargateur** [comment]
I got no idea how to remove git submodule change from a commit, I don't even know why it's show git submodule change when I didn't commit any submodule change... I hate git submodules so hard

**Comment by thomcc** [comment]
I'm going to be away for a few months, so I'm rerolling my PRs so that folks don't have to wait for me. Sorry/thanks.

r? libs

**Comment by ChrisDenton** [comment]
Hi @Stargateur, could you rebase this PR please?

File: library/std/src/sys/unix/time.rs [hunk 1]
```diff
@@ -70,12 +64,24 @@ impl fmt::Debug for SystemTime {
 
 impl Timespec {
     pub const fn zero() -> Timespec {
-        Timespec::new(0, 0)
+        unsafe { Self::new_unchecked(0, 0) }
     }
 
-    const fn new(tv_sec: i64, tv_nsec: i64) -> Timespec {
+    const fn new_assert(tv_sec: i64, tv_nsec: i64) -> Self {
```
Comment (ChrisDenton) [inline library/std/src/sys/unix/time.rs:70]: Could we not just replace this with `new().unwrap()`?

File: library/std/src/sys/unix/time.rs [hunk 2]
```diff
@@ -70,12 +64,24 @@ impl fmt::Debug for SystemTime {
...
         assert!(tv_nsec >= 0 && tv_nsec < NSEC_PER_SEC as i64);
         // SAFETY: The assert above checks tv_nsec is within the valid range
+        unsafe { Timespec::new_unchecked(tv_sec as i64, tv_nsec as i64) }
+    }
+
+    const fn new(tv_sec: i64, tv_nsec: i64) -> Result<Timespec, io::Error> {
+        if tv_nsec >= 0 && tv_nsec < NSEC_PER_SEC as i64 {
+            Ok(unsafe { Self::new_unchecked(tv_sec, tv_nsec) })
+        } else {
+            Err(io::const_io_error!(io::ErrorKind::Other, "Invalid time for nanosecond"))
```
Comment (m-ou-se) [inline library/std/src/sys/unix/time.rs:80]:
```suggestion
            Err(io::const_io_error!(io::ErrorKind::InvalidData, "Invalid timestamp"))
```

File: library/std/src/sys/unix/time.rs [hunk 3]
```diff
@@ -53,12 +53,6 @@ impl SystemTime {
     }
 }
 
-impl From<libc::timespec> for SystemTime {
```
Comment (riking) [inline library/std/src/sys/unix/time.rs:56]: Deleted From implementations on public types need to be at minimum replaced with a TryFrom implementation, or preferably not deleted at all

Comment (ChrisDenton) [reply library/std/src/sys/unix/time.rs:56]: Is `libc::timespec` public? I mean, the libc rust std uses.

Comment (Stargateur) [reply library/std/src/sys/unix/time.rs:56]: I removed it cause it's was not even used and was dangerous, AFAIK it's not public but I could be wrong.

File: library/std/src/sys/unix/time.rs (hunk 1 above)
Comment (Stargateur) [reply library/std/src/sys/unix/time.rs]: done

**Comment by ChrisDenton** [comment]
look good to me, thanks!

@bors r+

**Pull request merged**